src/geocel/vg/VecgeomTrackView.hh

```diff
@@ -397,7 +397,13 @@ CELER_FUNCTION bool VecgeomTrackView::is_on_boundary() const
  */
 CELER_FUNCTION Real3 VecgeomTrackView::normal() const
 {
-    CELER_NOT_IMPLEMENTED("VecgeomTrackView::normal");
+    CELER_EXPECT(this->is_on_boundary() || this->is_next_boundary());
+    vecgeom::Vector3D<vecgeom::Precision> normal;
+    auto pos = detail::to_vector(this->pos());
+    auto ok = this->is_on_boundary() ? vgstate_.Top()->Normal(pos, normal)
+                                     : vgnext_.Top()->Normal(pos, normal);
+    CELER_ENSURE(ok);
+    return Real3(normal);
```

Comment:
I think we can `CELER_EXPECT(!this->is_outside() && this->is_on_boundary())`

Also I'm surprised the last line works... it shouldn't! I see that Vector3D has an *implicit* `operator bool`, perhaps your vector is being converted to a bool which is being cast to a scalar that's filling the array's first element (i.e. returning {1,0,0} always). Try using `detail::to_array(normal)`.

Reply:
So in VecgeomTrackView::is_on_boundary() only checks vgstate_, not vgnext_, to be on the boundary.
I wanted to make sure that a track going from mother to daughter would be covered, as I would expect vgstate_ not to be on the boundary, in this case.